### ISSUE TITLE
recursive aggregation with partially existing aggregated variables

### ISSUE THREAD

**Issue by pjuergens** [IAMconsortium/pyam#525]
If I have some variables like 'Useful Energy|Transportation', 'Useful Energy|Transportation|Busses', 'Useful Energy|Transportation|Trucks' and 'Useful Energy|Space and Water Heating' and I do
```python
df.aggregate('Useful Energy', recursive=True, append=True)
```
I get an error because 'Useful Energy|Transportation' is already aggregated. I'd like to have the recursive aggregate function to recognise which variables already exist and skip aggregating them.

I didn't look at the source code yet. My naive idea would be something like
```python
if variable in df.variable:
    continue
else:
    #aggregate this variable
```

**Comment by danielhuppmann** [comment]
Thanks @pjuergens - not quite sure if what you're experiencing is a bug or a feature of pyam...

The recursive-aggregation feature is intended when you have data only at the lowest level of sectoral resolution and want to build up the variable tree hierarchy (currently only using summation).

If in your case, you already have data at an intermediate aggregation level, so there are a few questions:
1. Why do you already have data at an intermediate level? It already has been computed somewhere, so would it make sense to remove that earlier calculation and let the pyam-recursive-aggregation do its magic?
2. If you really need a more customized approach: why not explicitly call the `df.aggregate(..., recursive=False)` on the explicit hierarchy that you need?
3. If you still think that an implementation as you suggest is preferable, you should at least consider validating that the existing data at the intermediate level matches the aggregate of the lower-level components (see `check_aggregate()`) - otherwise, you risk having inconsistent values along the hierarchy. (But if you do the validation for every intermediate variable that already exists, there is no performance gain compared to just compute it all from the ground up).

Let's continue the discussion, or please close this issue if you come to the conclusion that your use case is not appropriate for the recursive-aggregation function.

**Comment by pjuergens** [comment]
By now I would call it a feature and not a bug. Still I'd like at least to have the option to use it the way I described. My use case would be a mixture of your points 1 and 3:

The output of the model I use (unfortunately) is a mixture of variables at the lowest level and intermediate aggregation levels. Sometimes the model might internally calculate aggregated values. Sometimes the values at the intermediate level are actually input data and the values at the lowest level are outputs. E.g. I could have Useful Energy as an Input, but Useful Energies at the lowest level as the optimized output of the model. In the end they should sum up to the Useful Energy-value at the highest (or intermediate) level.

So my wish was to aggregate everything with a single line of code and then use the check_aggregate-function to validate my output.

**Comment by Rlamboll** [comment]
I feel like modifying aggregate is better, maybe splitting the recursive option to accept a non-bool text argument like "checkIntermediate", which runs check_aggregate and throws an error if it fails, then throws away all the intermediate data and reconstructs it by running aggregate with recursive = True as before. This also prevents odd combinations like autocomplete = True, recursive = False.

**Comment by pjuergens** [comment]
That's basically what I do by now in my code as a workaround, just that I pick the intermediate variables by hand. So something like this, assuming I start with an IamDataFrame df:
```python
# define intermediate variables - this should be automated
intermediate_variables = ['Useful Energy|Transportation']
# throw away intermediate variables
_df = df.filter(variable=intermediate_variables, keep=False)
# aggregate resursively
_df.aggregate('Useful Energy', recursive=True, append=True)
# throw away variables that have been there before
_df = _df.filter(variable=df.variable, keep=False)
# add to old IamDataFrame
df.append(_df, inplace=True)
```

With check_aggregate I have the problem to differentiate between intermediate variables that are not followed directly by their components (e.g. I have 'Useful Energy|Transportation' and as lowest level 'Useful Energy|Transportation|LDV|BEV') and variables at the lowest level.

**Comment by danielhuppmann** [comment]
closing as implemented via #532
